core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalReader.java

```diff
@@ -11,101 +11,267 @@
+  /** Whether the reader runs in a primary master. */
+  private final boolean mPrimary;
+
+  /**
+   * The next edit log sequence number to read. This is not incremented when reading from
+   * the checkpoint.
+   */
+  private long mNextSequenceNumber;
+  /** The input stream to read the journal entries. */
+  private JournalInputStream mInputStream;
+  /** A queue of files to be processed including checkpoint and logs. */
+  private final Queue<UfsJournalFile> mFilesToProcess;
+  /** Buffer used to read from the file. */
+  private final byte[] mBuffer = new byte[1024];
+
+  /** Whether the reader is closed. */
+  private boolean mClosed;
+
+  /**
+   * A simple wrapper that wraps the journal file and the input stream.
+   */
+  private class JournalInputStream implements Closeable {
+    final UfsJournalFile mFile;
+    /** The input stream that reads from a file. */
+    final InputStream mStream;
+
+    JournalInputStream(UfsJournalFile file) throws IOException {
+      mFile = file;
+      mStream = mUfs.open(file.getLocation().toString());
+    }
+
+    /**
+     * @return whether we have finished reading the current file
+     */
+    boolean isDone() {
+      return mFile.getEnd() == mNextSequenceNumber;
+    }
+
+    @Override
+    public void close() throws IOException {
+      mStream.close();
+    }
+  }
 
   /**
    * Creates a new instance of {@link UfsJournalReader}.
    *
    * @param journal the handle to the journal
    */
-  UfsJournalReader(UfsJournal journal) {
+  UfsJournalReader(UfsJournal journal, JournalReaderOptions options) {
+    mFilesToProcess = new ArrayDeque<>();
     mJournal = Preconditions.checkNotNull(journal, "journal");
-    mUfs = UnderFileSystem.Factory.get(mJournal.getLocation().toString());
-    mCheckpoint = mJournal.getCheckpoint();
+    mUfs = mJournal.getUfs();
+    mNextSequenceNumber = options.getNextSequenceNumber();
+    mPrimary = options.isPrimary();
   }
 
   @Override
-  public boolean isValid() {
-    return mCheckpointRead && (mCheckpointOpenedTime == mCheckpointLastModifiedTime);
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+    if (mInputStream != null) {
+      mInputStream.close();
+    }
   }
 
   @Override
-  public JournalInputStream getCheckpointInputStream() throws IOException {
-    if (mCheckpointRead) {
-      throw new IOException("Checkpoint file has already been read.");
-    }
-    mCheckpointOpenedTime = getCheckpointLastModifiedTimeMs();
-
-    LOG.info("Opening journal checkpoint file: {}", mCheckpoint);
-    JournalInputStream jis =
-        mJournal.getJournalFormatter().deserialize(mUfs.open(mCheckpoint.toString()));
+  public long getNextSequenceNumberToCheckpoint() throws IOException {
+    return UfsJournalSnapshot.getNextLogSequenceToCheckpoint(mJournal);
+  }
 
-    mCheckpointRead = true;
-    return jis;
+  @Override
+  public long getNextSequenceNumber() {
+    return mNextSequenceNumber;
```

Comment:
It just tells the caller what is the next sequence number to read. I can start one reader, read one entry, get the next sequence number and pass that to another reader. These two readers can read in parallel.

But for all the callers we have, it is possible to force all the callers to only call this after the reader is closed.  I am ok if you want to enforce this just to be safer.